File: vijos/VIJOS数据和部分题解/P1518/Standard.cpp
#include <stdio.h>
int n,k;
int left[101],right[101],wi[101],di[101];
int f[101][101][51];
int min(int a,int b){
    return a<b?a:b;
}
void init(){
    int v,w,d,i;
    scanf("%d%d",&n,&k);
    for (i=0;i<=n;i++) left[i]=right[i]=n+1;
    for (v=0;v<=n;v++) for (w=0;w<=n;w++) for (d=0;d<=k;d++) f[v][w][d]=-1;
    for (i=1;i<=n;i++){
        scanf("%d%d%d",&w,&v,&d);
        right[i]=left[v];
	    left[v]=i;
	    di[i]=d;
	    wi[i]=w;
    }
}
int work(int from,int to,int k,int s,int tx){
    if(from==n+1||to==n+1) return 0;
    if(f[from][to][k]!=-1) return f[from][to][k];
    int x=2147483647,i;
    for (i=0;i<k;i++){
        x=min(x,work(left[from],from,i,0)+work(right[from],to,k-i-1,s));
    }
    for (i=0;i<=k;i++){
        x=min(x,work(left[from],to,i,s+di[from])+work(right[from],to,k-i,s)+(s+di[from])*wi[from]);
    }
    f[from][to][k]=x;
    return x;
}
int main(){
    init();
    printf("%d",work(left[0],0,k,0));
    return 0;
}

File: vijos/VIJOS数据和部分题解/P1524/Standard.pas
program Test;
var
  ga:array[1..100,1..100] of longint;
  n,ans:longint;

function min(a,b:longint):longint;
  begin
    if a<b then min:=a else min:=b;
  end;

function max_flow(s,t:longint):longint;
  var
    flow,pre:array[1..100] of longint;
    used:array[1..100] of boolean;
    max,max_loc,i,k,total:longint;
  begin
    total:=0;
    while true do
      begin
        fillchar(used,sizeof(used),false);
        fillchar(flow,sizeof(flow),0);
        fillchar(pre,sizeof(pre),0);
        flow[s]:=maxlongint;
        pre[s]:=-1;
        while true do
          begin
            max:=0;
            max_loc:=-1;
            for i:=1 to n+1 do
              if (not used[i]) and (flow[i]>max) then
                begin
                  max:=flow[i];
                  max_loc:=i;
                end;
            if (max_loc=-1) or (max_loc=t) then break;
            used[max_loc]:=true;
            for i:=1 to n+1 do
              if min(flow[max_loc],ga[max_loc,i])>flow[i] then
                begin
                  flow[i]:=min(flow[max_loc],ga[max_loc,i]);
                  pre[i]:=max_loc;
                end;
          end;
        if max_loc=-1 then
          exit(total);
        total:=total+flow[t];
        i:=t;
        while pre[i]<>-1 do
          begin
            k:=pre[i];
            dec(ga[k,i],flow[t]);
            inc(ga[i,k],flow[t]);
            i:=k;
          end;
      end;
  end;

procedure init;
  var
    i,a,b,e,m,x,w:longint;
  begin
    readln(n,e);
    fillchar(ga,sizeof(ga),0);
    for i:=1 to e do
      begin
        readln(a,b,w);
        ga[a,b]:=w;
        ga[b,a]:=w;
      end;
    readln(m);
    for i:=1 to m do
      begin
        read(x);
        ga[x,n+1]:=maxlongint;
      end;
  end;

procedure work;
  var
    i:longint;
  begin
    ans:=max_flow(1,n+1);
  end;

procedure print;
  begin
    writeln(ans);
  end;

begin
  init;
  work;
  print;
end.
